Maix Amigo IPS and Maix Bit Compatibility

odudex opened the pull request:
- Changed boot order to display splash while loading modules.
- Added several configurations and conditional codes to deal with different displays, cameras, buttons and touchscreen.
- Menus and keypads draw methods were modified to be compatible with touchscreen mapping.
- Selected items are now outlined.
- Changed font and alignment logic to use font width and height instead of font_size.
- Added a `touchscreen.py` file with driver and API to access Amigos touchscreen.
- QR animations are now `time` based to avoid different periods for different devices.

jreesun commented:
I have limited throughput the next few days, so my review might be delayed by up to a week, but this is still a high priority for me!

Thanks for your work on this, excited to see and try it out :)

jreesun commented:
Hoping to review this on the weekend

jreesun commented on src/krux/input.py:
```diff
@@ -42,35 +49,79 @@ def __init__(self):
+            else None
+        )
+
+    def get_button_c(self):
```
Two thoughts:
1) I think this pattern should be used for all buttons (i.e., `page` and `enter`)
2) The names of the methods ideally would match the underlying variable name, something like:  `page_state(self)`,  `enter_state(self)`, `previous_state(self)`, `touch_state(self)`

Also, try to avoid adding `get_` infront of the method names since it's not very "Pythonic". I've done it in a few places as well and need to change them...

jreesun commented on src/krux/input.py:
```diff
@@ -42,35 +49,79 @@ def __init__(self):
+            and self.get_button_c() == RELEASED
+            and self.get_touch_state() == TOUCH_IDLE
+        ):
+            wdt.feed() #here is were krux spend most of time
```
Need to increment `self.entropy` here too

jreesun commented on src/krux/input.py:
```diff
@@ -42,35 +49,79 @@ def __init__(self):
 
         if self.enter.value() == PRESSED:
             # Wait for release
             while self.enter.value() == PRESSED:
```
I think the `wdt.feed` should remain in these while loops, otherwise if a user held down a button press for > 30 seconds, the device would reset

jreesun commented on src/krux/pages/__init__.py:
```diff
@@ -48,6 +47,92 @@ def __init__(self, ctx, menu):
+        return DEFAULT_PADDING + self.ctx.display.font_height * 3
+
+    def _map_keys_array(self, width, height):
+        """Maps an arry of regions for keys to be placed in
```
typo, arry

jreesun commented on src/krux/pages/__init__.py:
```diff
@@ -48,6 +47,92 @@ def __init__(self, ctx, menu):
+
+    def _map_keys_array(self, width, height):
+        """Maps an arry of regions for keys to be placed in
+        Returns horizontal and vertican spacing of keys
```
typo, vertican

jreesun commented on src/krux/pages/__init__.py:
```diff
@@ -249,6 +339,31 @@ def print_qr_prompt(self, data, qr_format):
                 i += 1
 
+    def prompt(self, text, offset_y=0):
+        """Prompts user to answaer Yes or No"""
```
typo, answaer

jreesun commented on src/krux/pages/__init__.py:
```diff
@@ -235,7 +325,7 @@ def print_qr_prompt(self, data, qr_format):
         self.ctx.display.draw_centered_text(t("Print to QR?"))
         btn = self.ctx.input.wait_for_button()
-        if btn == BUTTON_ENTER:
+        if btn in (BUTTON_ENTER, BUTTON_TOUCH):
```
Can we tweak this so that there is a function in `input.py` where we can pass in a button and get back a boolean instead?

Something like:
```
def is_proceed_button(btn):
     return btn in (BUTTON_ENTER, BUTTON_TOUCH)
```
Then all these places could become:
```
if is_proceed_button(self.ctx.input.wait_for_button()):
```

This way we don't accidentally forget to put one of these buttons in a condition, with the added bonus of DRYing up the code.

jreesun commented on src/krux/touchscreen.py:
```diff
@@ -0,0 +1,151 @@
+            else:
+                self.index = 0
+
+    def get_state(self):
```
For symmetry with the other input buttons, could name this `value`

jreesun commented on src/krux/camera.py:
```diff
@@ -68,9 +74,10 @@ def capture_qr_code_loop(self, callback):
-
             res = img.find_qrcodes()
+            if board.config["type"] == "m5stickv":
+                img.lens_corr(strength=1.0, zoom=0.7)  # better fit the screen
```
Nice, thanks!

jreesun commented on src/krux/display.py:
```diff
@@ -240,34 +267,60 @@ def clear(self):
 
+    def draw_touch_button(
+        self, lines, offset_y=0
+    ):  # todo, lcd won't let draw rectangle over 320
```
Is this todo still true after your recent changes to `py_lcd`?

odudex replied:
Rectangle issue was already fixed on this PR, forgot to delete comment.
I'll go through all review items!
Next PR will also have a dedicated method for dynamic touch menu, removed draw_button and cleaned unused variables from `font` and `py_lcd`.
There are still some things to be adapted to touch, such as settings, but I think I should focus on settling all reviewed items and recent changes. Than, before any new code or feature, I could help with documentation, learn about and hopefully also help with tests and simulator.

jreesun reviewed (changes requested):
Great work! I left feedback on the things I would like to see changed before I merge this. 

The tests will need to be updated and new coverage added for the new code to keep total test coverage above 95%.

Also, we'll need to figure out a way to update the Github Pages docs site to mention the new devices supported.

Finally, this is mostly a note to myself, the simulator will need to be updated to include at least the Amigo.

odudex commented on src/krux/psbt.py:
```diff
@@ -71,7 +71,10 @@ def __init__(self, wallet, psbt_data):
         for inp in self.psbt.inputs:
             # get policy of the input
-            inp_policy = get_policy(inp, inp.witness_utxo.script_pubkey, xpubs)
+            try:
```
I used this when I was debugging and learning about Blue Wallet's legacy wallet creation issue when a xpub was given to it. As I remember, get_policy method couldn't deal with a legacy transaction read attempt and crashed. I think it is better to let it there until further tests, maybe give a more clear error message. Because if more people use Krux with Blue, there will be more people falling into this problem.

jreesun replied:
Can you file an issue for this? I would like to get to the bottom of it when I get time

jreesun replied:
Can you file an issue for this? I would like to get to the bottom of it when I get time

jreesun commented on src/krux/pages/login.py:
```diff
@@ -132,6 +130,8 @@ def _new_key_from_die(self, states, min_rolls, max_rolls):
                     roll = self.capture_from_keypad(
                         t("Roll %d") % (i + 1), states, lambda r: r
                     )
+                    if roll == "Esc":
```
This should be `display.ESC` so translated strings will also match (if different)

jreesun commented on src/krux/pages/login.py:
```diff
@@ -237,6 +234,8 @@ def _load_key_from_keypad(
                         autocomplete_fn,
                         possible_keys_fn,
                     )
+                    if word == "Esc":
```
This should be `display.ESC` so translated strings will also match (if different)

odudex replied:
My idea was not involve translations on this, there's no t() method here, but I will define a ESC_CODE constant to in the code

odudex replied:
Just used MENU_CONTINUE constant instead, it is the command that Esc executes and won't conflict with any key, worked fine.

jreesun commented on src/krux/display.py:
```diff
@@ -240,34 +266,36 @@ def clear(self):
-        lines_height = len(lines) * self.line_height()
-        offset_y = max(0, (screen_height - lines_height) // 2)
-        self.draw_hcentered_text(text, offset_y, color, padding)
+    def draw_centered_text(self, text, color=lcd.WHITE):
```
Can you add `bg_color=lcd.BLACK` to this as well?

jreesun commented on src/krux/display.py:
```diff
@@ -240,34 +266,36 @@ def clear(self):
+        self.draw_hcentered_text(text, offset_y, color)
 
-    def flash_text(self, text, color=lcd.WHITE, padding=DEFAULT_PADDING, duration=3000):
+    def flash_text(self, text, color=lcd.WHITE, duration=3000):
```
Can you add `bg_color=lcd.BLACK` to this as well (after color)?

odudex commented on krux:
```diff
@@ -72,18 +84,27 @@ elif [ "$1" == "flash" ]; then
     fi
 
-    docker run --privileged --device=/dev/$usb_device_name:/dev/ttyUSB0 --rm -w "$BUILD_DIR" -it krux-builder python3 ktool.py -B goE -p /dev/ttyUSB0 -b 1500000 kboot.kfpkg
+    docker run --privileged --device=/dev/$usb_device_name:/dev/ttyUSB0 --rm -w "$BUILD_DIR" -it krux-builder python3 ktool.py -B goE -p /dev/$usb_device_name -b 1500000 kboot.kfpkg
```
I tried that, but it won't work for Maix Amigo. Amigo creates ttyUSB0 and ttyUSB1, I think docker VM also creates both ports, and I can only flash Amigo if `-p` param specifies ttyUSB1. Bit also create both ports, but as it uses ttyUSB0 it works fine the way it is

jreesun replied:
Can you change this line to:
```
docker run --privileged --rm -w "$BUILD_DIR" -it krux-builder python3 ktool.py -B goE -p /dev/$usb_device_name -b 1500000 kboot.kfpkg
```

Does that resolve the issue for you?

odudex commented on krux:
```diff
@@ -34,7 +34,19 @@ getdevice() {
         if grep -q $idV $path/idVendor; then
             if grep -q $idP $path/idProduct; then
-                (find $path -name 'device' | rev | cut -d / -f 2 | rev) 2>/dev/null
+                (find $path -name 'device' | rev | cut -d / -f 2 | rev) 2>/dev/null | head -1
```
This worked! I just had to replace `num=${2:1}` for `num=${2}`

odudex replied:
I wonder if there is a way to try both ports, if it fails with `device_num=1` and get the "greeting fail" message, than tries with `device_num=2`. Because I don't know if the naming behavior I get here will repeat elsewhere.

jreesun commented:
@odudex It looks like `vendor/embit` was also updated. Can you revert back to commit `ac64b36818d9ee3f52b5a178966e3fb32136f7da`? I want to stick with latest tagged releases when possible

odudex commented:
Sorry, is it OK now?

jreesun commented on src/krux/touch.py:
```diff
@@ -0,0 +1,111 @@
+
+    def current_state(self):
+        """Returns the touchscreen state"""
+        if time.ticks_ms() > self.last_time + self.cycle:
```
Testing the build locally on my Amigo now and I noticed that you have to long-press to get button presses to work. I made a new build where I removed this timer condition (so `current_state` is now always checking when called) and it made touches more responsive. Could you make that change here? Both `self.cycle`  and `self.last_time` can be removed

odudex replied:
Before I do this: Did you let it running like this for a while? On my early tests with touch it would crash(IO error) if I didn't limit this I2C pulling frequency, maybe there was another cofactor on this issue that was solved later, but I think it is a good idea to let it running for some minutes, keep navigating on menus, run some QR scans, etc. to check if it remains stable.

odudex replied:
I tested here, and after 5 min running it happened:
```
Traceback (most recent call last):                                                                                                
  File "boot.py", line 42, in <module>                                                                                            
  File "krux/pages/__init__.py", line 315, in run                                                                                 
  File "krux/pages/__init__.py", line 334, in run_loop                                                                            
  File "krux/input.py", line 80, in wait_for_button                                                                               
  File "krux/input.py", line 76, in wait_for_press                                                                                
  File "krux/input.py", line 58, in touch_value                                                                                   
  File "krux/touch.py", line 70, in value                                                                                         
  File "krux/touch.py", line 60, in current_state                                                                                 
  File "krux/touchscreens/ft6x36.py", line 30, in current_point                                                                   
  File "krux/touchscreens/ft6x36.py", line 26, in read_reg                                                                        
OSError: [Errno 5] EIO                                                                                                            
MicroPython v1.11 on 2022-04-21; Sipeed_M1 with kendryte-k210                                                                     
Type "help()" for more information.        
```
Maybe there are ways to prevent this, like checking for a busy I2C, but I don't know exactly what is causing this, for now I would let it the way it is, maybe the sample rate can be increased a bit, but I would test well before doing it.

jreesun replied:
Strange, I actually used it for a while (much longer than 5 minutes) and didn't run into this error. Nonetheless, if it's an intermittent error, maybe we can just keep track of the last point successfully read from the register and wrap reads in `current_point` in a try..catch:

```
    def current_point(self):
        """If touch is pressed, returns x and y points"""
        try:
            data = self.read_reg(0x02, 1)
            if data is not None and data[0] == 0x1:
                data_buf = self.read_reg(0x03, 4)
                x = ((data_buf[0] & 0x0F) << 8) | (data_buf[1])
                y = ((data_buf[2] & 0x0F) << 8) | (data_buf[3])
                if (data_buf[0] & 0xC0) == 0x80:
                    self.last_point = (x, y)
                    return self.last_point
        except:
            return self.last_point

        self.last_point = None
        return self.last_point
```
and in  `__init__`:

`self.last_point = None`

Can you give that a try and see if it fixes your problem?

jreesun commented on src/krux/touch.py:
```diff
@@ -0,0 +1,111 @@
+
+    def current_state(self):
+        """Returns the touchscreen state"""
+        if time.ticks_ms() > self.last_time + self.cycle:
```
Testing the build locally on my Amigo now and I noticed that you have to long-press to get button presses to work. I made a new build where I removed this timer condition (so `current_state` is now always checking when called) and it made touches more responsive. Could you make that change here? Both `self.cycle`  and `self.last_time` can be removed

jreesun commented:
Does the flash light work for you when reading a QR code and pressing Enter/BUTTON_A? It's not working on my device, but I'm not sure why it wouldn't be... the pin for LED_W seems correct to me. Not sure if it's just my device or not.

jreesun commented on src/krux/touchscreens/ft6x36.py:
```diff
@@ -0,0 +1,71 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
```
Adding `# pylint: disable=W0231` here will fix the pylint error

odudex replied:
Replaying all here: 
- I'll test the `try` and see if it solves, there's no need to store last position, I'll just discard an eventual erroneous reading.
- Yes, the LED works fine on my Amigo

odudex commented:
Ok, I will change the order and prompt on Esc

odudex commented:
During tests on `try - except`  on I2C touch pulling I started to get errors (not crashes) from PMU's pulls from I2C.
I'll investigate if there is a better and neat solution for this. But I think keeping the I2C bus busy all time may not be a good idea.

jreesun commented:
> During tests on `try - except` on I2C touch pulling I started to get errors (not crashes) from PMU's pulls from I2C. I'll investigate if there is a better and neat solution for this. But I think keeping the I2C bus busy all time may not be a good idea.

Okay, you can leave what you had before then. Thanks for giving it a try!

jreesun commented on src/krux/pages/__init__.py:
```diff
@@ -195,15 +289,14 @@ def display_qr_codes(
-            done = btn == BUTTON_ENTER
-            if not done:
-                time.sleep_ms(QR_ANIMATION_INTERVAL_MS)
+            if self.ctx.input.wait_for_proceed():
```
`wait_for_proceed` needs to be updated to take an optional `block=True` param like `wait_for_button`,  and the same expression should be passed to it as was here before:
`if self.ctx.input.wait_for_proceed(block=num_parts == 1)`

Otherwise, QR codes will stop animating automatically

jreesun commented:
After going through and testing the device, here are my remaining nits:
- [necessary] The "Sign?" prompt for signing a message doesn't have Yes/No buttons to touch
- [necessary] Settings should be navigable by touching (new) `<` and  `>` buttons, and touching other parts of the screen on a setting should act like the enter button
- [necessary] QR codes aren't animating (automatically) anymore. See my other comment above
- [nice-to-have] buttons_active should be reset every time the active "page" (menu) changes
- [nice-to-have] highlight box around buttons doesn’t go full-width like button dividers do; looks odd
- [nice-to-have] yes/no buttons should draw at the exact bottom of screen without leaving a vertical gap/padding
- [nice-to-have] yes/no buttons need some padding between them and the prompt text
- [nice-to-have] header/title of pages need some padding between them and content of page (e.g., keypad input pages, mnemonic page)
- [nice-to-have] remove double-lines on keypad inputs… shared lines should overlap and look like a single line

After those are addressed, I think this is ready for test coverage, then merging.

Sorry to be such a pain; I really appreciate all your effort on this!

jreesun reviewed (commented):
After going through and testing the device, here are my remaining nits:
- [necessary] The "Sign?" prompt for signing a message doesn't have Yes/No buttons to touch
- [necessary] Settings should be navigable by touching (new) `<` and  `>` buttons, and touching other parts of the screen on a setting should act like the enter button
- [necessary] QR codes aren't animating (automatically) anymore. See my other comment above
- [nice-to-have] buttons_active should be reset every time the active "page" (menu) changes
- [nice-to-have] highlight box around buttons doesn’t go full-width like button dividers do; looks odd
- [nice-to-have] yes/no buttons should draw at the exact bottom of screen without leaving a vertical gap/padding
- [nice-to-have] yes/no buttons need some padding between them and the prompt text
- [nice-to-have] header/title of pages need some padding between them and content of page (e.g., keypad input pages, mnemonic page)
- [nice-to-have] remove double-lines on keypad inputs… shared lines should overlap and look like a single line

After those are addressed, I think this is ready for test coverage, then merging.

Sorry to be such a pain; I really appreciate all your effort on this!

odudex commented:
- Dug a little deeper on I2C issue and I think issue is solved.
pmu had a timed i2C pull that checked if power button was being pressed. It is used to wake from sleep mode when device is turned off. And when it was on, it would reset the device if the power button was pressed.
I believe this timed I2C write/read could be triggered during a touch reading and create a conflict.
Since Amigo already have a dedicated reset button I disabled this timed I2C pwr button check when device is on, and enable it just before device is turned off to keep the "wake up" functionality. I also kept the try-except on touch read just in case.
Now touch is faster and I got no errors so far.
- Button order was changed, but there is small issue now: Clr and Esc buttons get too close from each other on M5StickV. Their position use same logic of touch keypads, with fixed width size buttons.
- QR Animation fixed
- I'm not familiarized with message signing, could you test it? I put the Y/N prompt on it
- Adjusted Y/N prompt paddings
- Implemented PAGE_PREV button on settings
- Touch buttons were added on settings, slightly different than you asked for, I also added a "Back" button. It would require a change in index algorithm to accept the "click elsewhere" command
- Odd appearance on menus + outlines were fixed. Now when touch is used outlines disappear, and when physical are used touch regions disappear

jreesun commented:
> Button order was changed, but there is small issue now: Clr and Esc buttons get too close from each other on M5StickV. Their position use same logic of touch keypads, with fixed width size buttons.

Getting rid of `Clr` would be the easiest way to solve this. I think it's redundant anyway since `<` can now remove chars without autocomplete interfering.

> I'm not familiarized with message signing, could you test it? I put the Y/N prompt on it

I tested and it works, thanks! By the way, it's easy to test, you can scan virtually any QR code and it will hash the contents and prompt you to sign them.

> Touch buttons were added on settings, slightly different than you asked for, I also added a "Back" button. It would require a change in index algorithm to accept the "click elsewhere" command

👌 

> Odd appearance on menus + outlines were fixed. Now when touch is used outlines disappear, and when physical are used touch regions disappear

👌

jreesun commented:
@odudex Going to merge this into `develop` now so I can push up some of my own fixes afterward. Can you make a new PR after this with test coverage?

jreesun commented:
@odudex I merged my follow-up PR (#126) just now which:
- cleaned up the black and pylint errors
- added missing translations
- fixed the simulator for the M5StickV (still need to add the Amigo to it though)

The remaining items to do are:
- Add test coverage for the new code (Krux has >95% code coverage and I want to keep it that way) and fix existing tests that were broken by changes this introduced
- Update the README to mention the devices Krux supports

I need to step back and work on other things now, so I'm going to let you work on the test coverage. @ me if you hit any snags or run out of time, I'll see if I can help.

Once coverage is back to 95% and all tests are passing, I'll merge `develop` into `main` and put out a new release with signed builds for m5stickv, amigo, and bit.

Finally, we'll need to update the Github Pages site (under `docs`) to mention the newly-supported devices, ideally with screenshots from the simulator. I'll file an issue for this since the changes to the site could end up being pretty extensive, but I don't think it's necessary to have this done before the release.

Once again, thank you so much for your contributions here! You did a fantastic job on this; Krux on the Amigo is really cool!

odudex commented:
@jreesun nice you set a milestone on this by creating `develop` branch.
My "regular job" is still not demanding too much from now so I intend to keep going on Krux next week, if scenario changes I'll let you know.
Don't worry about "being a pain", all things you pointed would have to be done sooner or later.
Nice you a have an Amigo now too. Krux indeed is really cool on it!
